File: libs/ui_wellseis/src/qt/cross_well_export_dialog.cpp
#include <pwb/ui_wellseis/qt/cross_well_export_dialog.hpp>

#include <QComboBox>
#include <QDialogButtonBox>
#include <QFormLayout>
#include <QSpinBox>
#include <QVBoxLayout>

namespace pwb::ui_wellseis::qt {

namespace {

QString qs(const std::string& text) {
    return QString::fromUtf8(text.data(), static_cast<int>(text.size()));
}

}  // namespace

CrossWellExportDialog::CrossWellExportDialog(QWidget* parent)
    : QDialog(parent) {
    setWindowTitle(QStringLiteral("导出连井剖面"));
    auto* layout = new QVBoxLayout(this);
    auto* form = new QFormLayout();

    fmt_combo_ = new QComboBox(this);
    fmt_combo_->addItem(QStringLiteral("SVG 矢量图"), QStringLiteral("svg"));
    fmt_combo_->addItem(QStringLiteral("PNG 位图"), QStringLiteral("png"));
    fmt_combo_->addItem(QStringLiteral("PDF 文档"), QStringLiteral("pdf"));
    connect(fmt_combo_, &QComboBox::currentIndexChanged, this,
            [this] { update_enabled(); });
    form->addRow(QStringLiteral("格式"), fmt_combo_);

    dpi_combo_ = new QComboBox(this);
    for (const int dpi : {96, 150, 300}) {
        dpi_combo_->addItem(QString::number(dpi), dpi);
    }
    dpi_combo_->setCurrentIndex(1);  // 150
    form->addRow(QStringLiteral("DPI"), dpi_combo_);

    width_spin_ = new QSpinBox(this);
    width_spin_->setRange(0, 20000);
    width_spin_->setSingleStep(100);
    width_spin_->setSpecialValueText(QStringLiteral("自然宽度"));
    width_spin_->setSuffix(QStringLiteral(" px"));
    width_spin_->setValue(0);
    form->addRow(QStringLiteral("宽度"), width_spin_);

    page_size_combo_ = new QComboBox(this);
    page_size_combo_->addItem(QStringLiteral("内容尺寸"), QVariant{});
    page_size_combo_->addItem(QStringLiteral("A4"), QStringLiteral("A4"));
    page_size_combo_->addItem(QStringLiteral("Letter"),
                              QStringLiteral("LETTER"));
    connect(page_size_combo_, &QComboBox::currentIndexChanged, this,
            [this] { update_enabled(); });
    form->addRow(QStringLiteral("纸张"), page_size_combo_);

    layout->addLayout(form);
    auto* buttons = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel, this);
    connect(buttons, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(buttons, &QDialogButtonBox::rejected, this, &QDialog::reject);
    layout->addWidget(buttons);

    update_enabled();
}

CrossWellExportOptions CrossWellExportDialog::raw_options() const {
    CrossWellExportOptions options;
    options.fmt = fmt_combo_->currentData().toString().toStdString();
    options.dpi = dpi_combo_->currentData().toInt();
    options.width_px = width_spin_->value();
    const QVariant page = page_size_combo_->currentData();
    if (page.isValid() && !page.toString().isEmpty()) {
        options.page_size = page.toString().toStdString();
    }
    return options;
}

ResolvedExportOptions CrossWellExportDialog::options() const {
    return resolve_export_options(raw_options());
}

void CrossWellExportDialog::set_options(
    const CrossWellExportOptions& options) {
    const int fmt_index = fmt_combo_->findData(qs(options.fmt));
    if (fmt_index >= 0) {
        fmt_combo_->setCurrentIndex(fmt_index);
    }
    const int dpi_index = dpi_combo_->findData(options.dpi);
    if (dpi_index >= 0) {
        dpi_combo_->setCurrentIndex(dpi_index);
    }
    width_spin_->setValue(options.width_px);
    if (options.page_size.has_value()) {
        const int page_index =
            page_size_combo_->findData(qs(*options.page_size));
        if (page_index >= 0) {
            page_size_combo_->setCurrentIndex(page_index);
        }
    } else {
        page_size_combo_->setCurrentIndex(0);
    }
    update_enabled();
}

void CrossWellExportDialog::update_enabled() {
    const CrossWellExportEnabled enabled = export_enabled(raw_options());
    dpi_combo_->setEnabled(enabled.dpi);
    page_size_combo_->setEnabled(enabled.page_size);
    width_spin_->setEnabled(enabled.width);
}

}  // namespace pwb::ui_wellseis::qt

File: legacy/python_reference/product/paleo_workbench/ui/pages/cross_well_export_dialog.py
"""Export options dialog for the cross-well correlation section."""

from __future__ import annotations

from PySide6.QtWidgets import (
    QComboBox,
    QDialog,
    QDialogButtonBox,
    QFormLayout,
    QSpinBox,
    QVBoxLayout,
    QWidget,
)


class CrossWellExportDialog(QDialog):
    """Pick export format, DPI, width and PDF page size."""

    def __init__(self, parent: QWidget | None = None):
        super().__init__(parent)
        self.setWindowTitle("导出连井剖面")
        layout = QVBoxLayout(self)
        form = QFormLayout()

        self.format_combo = QComboBox()
        self.format_combo.addItem("SVG 矢量图", "svg")
        self.format_combo.addItem("PNG 位图", "png")
        self.format_combo.addItem("PDF 文档", "pdf")
        self.format_combo.currentIndexChanged.connect(self._update_enabled)
        form.addRow("格式", self.format_combo)

        self.dpi_combo = QComboBox()
        for dpi in (96, 150, 300):
            self.dpi_combo.addItem(str(dpi), dpi)
        self.dpi_combo.setCurrentIndex(1)  # 150
        form.addRow("DPI", self.dpi_combo)

        self.width_spin = QSpinBox()
        self.width_spin.setRange(0, 20000)
        self.width_spin.setSingleStep(100)
        self.width_spin.setSpecialValueText("自然宽度")
        self.width_spin.setSuffix(" px")
        self.width_spin.setValue(0)
        form.addRow("宽度", self.width_spin)

        self.page_size_combo = QComboBox()
        self.page_size_combo.addItem("内容尺寸", None)
        self.page_size_combo.addItem("A4", "A4")
        self.page_size_combo.addItem("Letter", "LETTER")
        self.page_size_combo.currentIndexChanged.connect(self._update_enabled)
        form.addRow("纸张", self.page_size_combo)

        layout.addLayout(form)
        buttons = QDialogButtonBox(
            QDialogButtonBox.StandardButton.Ok | QDialogButtonBox.StandardButton.Cancel
        )
        buttons.accepted.connect(self.accept)
        buttons.rejected.connect(self.reject)
        layout.addWidget(buttons)

        self._update_enabled()

    def _update_enabled(self) -> None:
        fmt = self.format_combo.currentData()
        self.dpi_combo.setEnabled(fmt in ("png", "pdf"))
        self.page_size_combo.setEnabled(fmt == "pdf")
        self.width_spin.setEnabled(
            not (fmt == "pdf" and self.page_size_combo.currentData() is not None)
        )

    def options(self) -> dict:
        fmt = self.format_combo.currentData()
        page_size = self.page_size_combo.currentData() if fmt == "pdf" else None
        width_px = self.width_spin.value() or None
        if page_size is not None:
            width_px = None
        return {
            "fmt": fmt,
            "dpi": self.dpi_combo.currentData(),
            "width_px": width_px,
            "page_size": page_size,
        }
